src/apps/hidden-hand/common/hidden-hand.rewards.contract-position-fetcher.ts

```diff
@@ -0,0 +1,138 @@
+import { MetaType } from '~position/position.interface';
+import { ZERO_ADDRESS } from '~app-toolkit/constants/address';
+import { NETWORK_IDS } from '~types';
+
+export class HiddenHandRewardsContractPositionFetcher extends CustomContractPositionTemplatePositionFetcher<
+  HiddenHandRewardDistributor,
+  DefaultDataProps,
+  HiddenHandRewardsDefinition
+> {
+  groupLabel: string = 'Rewards';
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(HiddenHandContractFactory) protected readonly contractFactory: HiddenHandContractFactory,
+    @Inject(HiddenHandRewardsResolver) protected readonly rewardsResolver: HiddenHandRewardsResolver,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string): HiddenHandRewardDistributor {
+    const a_ = REWARD_DISTRIBUTOR[this.network] || '';
+    return this.contractFactory.hiddenHandRewardDistributor({ address: a_, network: this.network });
+  }
+
+  async getDefinitions() {
+    const networkProtocols = PROTOCOLS[this.network] || {};
+    const definitions = Object.entries(networkProtocols).map(([key, protocol]) => ({
+      address: protocol.address,
+      name: protocol.name,
+      key,
+    }));
+
+    return definitions;
+  }
+
+  async getTokenDefinitions(
+    { address, contract }: GetTokenDefinitionsParams<Contract, DefaultContractPositionDefinition>,
+  ): Promise<UnderlyingTokenDefinition[] | null> {
+    const distribution = await this.rewardsResolver.getData();
+    const protocols = PROTOCOLS[this.network] || {};
+    const neworkId = NETWORK_IDS[this.network] || '';
+    const vault = (await contract.BRIBE_VAULT()).toLowerCase();
+
+    const marketEntry = Object.entries(protocols).find(
+      ([, protocolInfo]) => protocolInfo.address.toLowerCase() === address.toLowerCase()
+    );
+    const market = marketEntry ? marketEntry[0] : '';
+
+    if (!distribution?.hasOwnProperty(market)) return [];
+
+    const claimableDistribution = distribution[market][neworkId] ?? {};
+    const claimableTokens = Object.keys(claimableDistribution);
+
+    const baseTokens = await this.appToolkit.getBaseTokenPrices(this.network);
+
+    const tokenDefinitions = claimableTokens.map(token => {
+      const token_ = token.toLowerCase() === vault ? ZERO_ADDRESS : token.toLowerCase();
+      const tokenFound = baseTokens.find(p => p.address === token_);
+      if (!tokenFound) return null;
+      return {
+        metaType: MetaType.CLAIMABLE,
+        address: token_,
+        network: this.network,
+      }
+    });
+
+    return _.compact(tokenDefinitions);
+  }
+
+  async getLabel(
+    { definition }: GetDisplayPropsParams<Contract, DefaultDataProps, HiddenHandRewardsDefinition>,
+  ): Promise<string> {
+    return definition.name;
+  }
+
+  getTokenBalancesPerPosition(): never {
+    throw new NotImplementedException();
+  }
+
+  async getBalances(address: string): Promise<ContractPositionBalance<DefaultDataProps>[]> {
```

Comment:
Would it be possible to fetch the balance making calls on chain ?

Reply:
Unfortunately not. Only data on chain is the claimed balance. I'm fetching the distribution and subtracting the claimed amount in `getClaimableDistribution()`